coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/functions/CoralSqlUnnestOperator.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2022 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.coral.hive.hive2rel.functions;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.SqlUnnestOperator;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.type.ArraySqlType;
+import org.apache.calcite.sql.type.MapSqlType;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+
+/**
+ * This class is Coral's representation for Hive explode function.
+ * Explode enables unboxing an array or map of an operand.
+ * The function returns a row set of single column for the array operand, or
+ * a row set with two columns corresponding to (key, value) for
+ * map operand type. The array operand may be a primitive datatype, such as int, String.
+ * Or, it may also be an abstract datatype, such as struct.
+ */
+public class CoralSqlUnnestOperator extends SqlUnnestOperator {
+
+  // _withOrdinality represents whether output should contain an ORDINALITY column
+  final boolean _withOrdinality;
+  // _sqlDialect represents the target language's dialect. By default, it is set to spark
+  String _sqlDialect = "spark";
+  // _relDataType represents the datatype of the operand to be unboxed.
+  RelDataType _relDataType;
+
+  public static final String ARRAY_ELEMENT_COLUMN_NAME = "col";
+  public static final String ARRAY_ELEMENT_POS_NAME = "pos";
+
+  public CoralSqlUnnestOperator(boolean withOrdinality, RelDataType relDataType) {
+    // keep the same as base class 'UNNEST' operator
+    // Hive has a separate 'posexplode' function for ordinality
+    super(withOrdinality);
+
+    _withOrdinality = withOrdinality;
+    _relDataType = relDataType;
+  }
+
+  public void setSqlDialect(String sqlDialect) {
+    _sqlDialect = sqlDialect;
+  }
+
+  public RelDataType getRelDataType() {
+    return _relDataType;
+  }
+
+  @Override
+  public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFailure) {
```

Comment:
Good point, looks like we dont necessarily need to override this method. The super seems to be initializing a simple generic [SqlOperandTypeChecker](https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java#L57) which will do the job.
Our implementation is just clearer and checks for possible permutations with ordinality as well. Do you think its an overkill and can be removed?
If its needed, we can extend `SqlFunctionalOperator` instead of `SqlUnnestOperator` which will allow us to customize the `SqlOperandTypeChecker`.

Reply:
IIUC, this class is to combine the original `HiveExplodeOperator` and `HivePosExplodeOperator`? 
If so, I think it makes sense to override `checkOperandTypes` to ensure input type to be map/array.

Reply:
@wmoustafa The parent's implementation involves instantiating an `operandTypeChecker` [here](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/SqlUnnestOperator.java#L55). This enables the super to impose checks on the operand types and make sure that operands are [one of the following](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java#L581):

1. family(SqlTypeFamily.MULTISET)
2. family(SqlTypeFamily.ARRAY) 
3. family(SqlTypeFamily.MAP)
4. [OperandTypes.RECORD_COLLECTION](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java#L517) 

CoralSqlUnnestOperator has these checks available by default. By overriding the function `checkOperandTypes(..)`, we introduced some 'good to have' additional checks.

Reply:
The 'good to have' override checks if a permutation of ordinality + operandType is allowed.
For instance, looking at the legacy implementation, we don't allow [ordinality + map].